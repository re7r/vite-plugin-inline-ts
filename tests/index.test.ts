import { afterEach, beforeEach, describe, expect, it } from 'vitest';
import { build, type Plugin } from 'vite';
import path from 'node:path';
import fs from 'node:fs';
import inlineTs from '../dist/index.js';

const outputDir = path.resolve(__dirname, 'output');
const fixturesDir = path.resolve(__dirname, 'fixtures');

const captureAssets: Record<string, string> = {};
const captureHtmlPlugin: Plugin = {
  name: 'capture-html',

  transform(code: string, id: string) {
    if (id.endsWith('.html') || id.endsWith('.xht')) {
      captureAssets[path.basename(id)] = code;
      return '';
    }
  },

  generateBundle() {
    for (const [fileName, source] of Object.entries(captureAssets)) {
      this.emitFile({
        type: 'asset',
        fileName,
        source,
      });
    }
  },
};

const engines = ['oxc', 'swc', 'esbuild', 'typescript'] as const;

describe.each(engines)('inlineTs engine: %s', (engine) => {
  beforeEach(() => {
    fs.rmSync(outputDir, { recursive: true, force: true });
    fs.mkdirSync(outputDir, { recursive: true });
  });

  afterEach(() => {
    fs.rmSync(outputDir, { recursive: true, force: true });
  });

  it(`processes HTML with engine=${engine}`, async () => {
    const inputFile = path.join(fixturesDir, 'main.mjs');
    await build({
      logLevel: 'silent',
      build: {
        rollupOptions: {
          input: inputFile,
        },
        outDir: outputDir,
      },
      plugins: [
        inlineTs({
          engine,
          extensions: ['.html', '.xht'],
          debug: true,
        }),
        captureHtmlPlugin,
      ],
    });

    fs.rmSync(path.join(outputDir, 'tests'), { recursive: true, force: true });
    fs.rmSync(path.join(outputDir, 'assets'), { recursive: true, force: true });

    const contents: Record<string, string> = {};
    const files = fs.readdirSync(outputDir).sort();
    expect(files).toMatchSnapshot(`${engine} file list`);

    for (const file of files) {
      const fullPath = path.join(outputDir, file);
      contents[file] = fs.readFileSync(fullPath, 'utf8');
    }

    expect(contents).toMatchSnapshot(`${engine} file contents`);
  });
});
