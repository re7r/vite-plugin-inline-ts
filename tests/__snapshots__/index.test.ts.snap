// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`inlineTs engine: esbuild > processes HTML with engine=esbuild > esbuild file contents 1`] = `
{
  "example.html": "<html>
  <head>
    <script>
      const v = 'var';
      console.log('native-html-js', v);
    </script>
    <script >
const s = "str";
console.log("compiled-html-js", s);

</script>
  </head>
  <body>
    Hello

    <script >
const s1 = "str-1";
console.log("compiled-html-js-1", s);

</script>
  </body>
</html>
",
  "example.xht": "<script >
const s = "hello";
console.log("compiled-xht-js", s);

</script>

<div>Hello</div>
",
}
`;

exports[`inlineTs engine: esbuild > processes HTML with engine=esbuild > esbuild file list 1`] = `
[
  "example.html",
  "example.xht",
]
`;

exports[`inlineTs engine: oxc > processes HTML with engine=oxc > oxc file contents 1`] = `
{
  "example.html": "<html>
  <head>
    <script>
      const v = 'var';
      console.log('native-html-js', v);
    </script>
    <script >
const s = "str";
console.log("compiled-html-js", s);

</script>
  </head>
  <body>
    Hello

    <script >
const s1 = "str-1";
console.log("compiled-html-js-1", s);

</script>
  </body>
</html>
",
  "example.xht": "<script >
const s = "hello";
console.log("compiled-xht-js", s);

</script>

<div>Hello</div>
",
}
`;

exports[`inlineTs engine: oxc > processes HTML with engine=oxc > oxc file list 1`] = `
[
  "example.html",
  "example.xht",
]
`;

exports[`inlineTs engine: swc > processes HTML with engine=swc > swc file contents 1`] = `
{
  "example.html": "<html>
  <head>
    <script>
      const v = 'var';
      console.log('native-html-js', v);
    </script>
    <script >
var s = 'str';
console.log('compiled-html-js', s);

</script>
  </head>
  <body>
    Hello

    <script >
var s1 = 'str-1';
console.log('compiled-html-js-1', s);

</script>
  </body>
</html>
",
  "example.xht": "<script >
var s = 'hello';
console.log('compiled-xht-js', s);

</script>

<div>Hello</div>
",
}
`;

exports[`inlineTs engine: swc > processes HTML with engine=swc > swc file list 1`] = `
[
  "example.html",
  "example.xht",
]
`;

exports[`inlineTs engine: typescript > processes HTML with engine=typescript > typescript file contents 1`] = `
{
  "example.html": "<html>
  <head>
    <script>
      const v = 'var';
      console.log('native-html-js', v);
    </script>
    <script >
const s = 'str';
console.log('compiled-html-js', s);

</script>
  </head>
  <body>
    Hello

    <script >
const s1 = 'str-1';
console.log('compiled-html-js-1', s);

</script>
  </body>
</html>
",
  "example.xht": "<script >
const s = 'hello';
console.log('compiled-xht-js', s);

</script>

<div>Hello</div>
",
}
`;

exports[`inlineTs engine: typescript > processes HTML with engine=typescript > typescript file list 1`] = `
[
  "example.html",
  "example.xht",
]
`;
